File: lib/qwen.js
// Qwen 视觉分析：调用阿里云百炼 qwen-vl，输出服装检测 + 标签（无 box）
// 无 API Key 时进入 demo 模式，返回合理 mock 结果（仅用于演示完整流程）
// 说明：本方案不要求模型输出包围盒；服饰分割交由 qwen-image 在原图上完成。
import fs from 'node:fs';
import { CONFIG } from './config.js';
import { SYSTEM_PROMPT, buildUserMessage } from './prompt.js';

// 从模型可能夹带的文本中提取 JSON 数组
function extractJSON(text) {
  if (!text) return [];
  let t = text.trim();
  // 去掉 ```json ... ``` 包裹
  const fence = t.match(/```(?:json)?\s*([\s\S]*?)```/i);
  if (fence) t = fence[1].trim();
  // 截取第一个 [ 到最后一个 ]
  const s = t.indexOf('[');
  const e = t.lastIndexOf(']');
  if (s === -1 || e === -1 || e < s) return [];
  try {
    return JSON.parse(t.slice(s, e + 1));
  } catch {
    return [];
  }
}

// 四大类（唯一合法类别，内部用英文做校验）
const VALID_CATEGORIES = ['Top', 'Bottom', 'Shoes', 'Bag'];

// ---------- 中文字典：把模型返回的英文受控词统一转换为中文存储 ----------
const CATEGORY_CN = { Top: '上装', Bottom: '下装', Shoes: '鞋', Bag: '包' };
const COLOR_CN = {
  White: '白色', Black: '黑色', Blue: '蓝色', Beige: '米色', Red: '红色',
  Green: '绿色', Gray: '灰色', Grey: '灰色', Pink: '粉色', Brown: '棕色',
  Yellow: '黄色', Purple: '紫色', Multicolor: '多色', Unknown: '未知',
};
const SEASON_CN = {
  Spring: '春季', Summer: '夏季', Autumn: '秋季', Fall: '秋季',
  Winter: '冬季', All: '四季', Unknown: '四季',
};
const MATERIAL_CN = {
  Cotton: '棉', Denim: '牛仔布', Leather: '皮革', Knit: '针织',
  Polyester: '聚酯纤维', Silk: '丝绸', Linen: '亚麻', Wool: '羊毛', Unknown: '未知',
};
const STYLE_CN = {
  Minimal: '极简', Casual: '休闲', Sporty: '运动', Formal: '正装', Vintage: '复古',
  Y2K: 'Y2K', Streetwear: '街头', French: '法式', OldMoney: '老钱风',
  Preppy: '学院风', Athleisure: '运动休闲', Unknown: '未知',
};
const FIT_CN = { Slim: '修身', Regular: '常规', Loose: '宽松', Oversized: '超大', Unknown: '未知' };
const PATTERN_CN = {
  Solid: '纯色', Striped: '条纹', Floral: '碎花', Plaid: '格纹',
  Dotted: '波点', Graphic: '图案', Other: '其它', Unknown: '其它',
};

// 英文名 → 中文；未知/异常时回退
function toCn(map, v, fallback = '未知') {
  const key = String(v || '').trim();
  return map[key] || (key ? key : fallback);
}

// 类别别名 → 规范类别。
// 连衣裙/外套/大衣/西装/夹克 统一归入上装(Top)；其余（含配饰）一律丢弃。
const CATEGORY_ALIASES = {
  top: 'Top', '上装': 'Top', '上衣': 'Top', '上裝': 'Top',
  bottom: 'Bottom', '下装': 'Bottom', '裤子': 'Bottom', '裤': 'Bottom', '下裝': 'Bottom',
  shoes: 'Shoes', '鞋': 'Shoes', '鞋子': 'Shoes', '鞋类': 'Shoes',
  bag: 'Bag', '包': 'Bag', '包包': 'Bag', '包类': 'Bag',
  // 归入上装
  dress: 'Top', '连衣裙': 'Top', '连身裙': 'Top',
  outerwear: 'Top', '外套': 'Top', '大衣': 'Top', 'coat': 'Top', 'jacket': 'Top', '夹克': 'Top',
  suit: 'Top', '西装': 'Top', '西裝': 'Top',
};

// 将模型给出的类别归一为规范类别；无法识别（如配饰）返回 null（将被丢弃）
function canonicalCategory(raw) {
  const c = String(raw || '').trim();
  if (!c) return null;
  const lower = c.toLowerCase();
  if (CATEGORY_ALIASES[lower]) return CATEGORY_ALIASES[lower];
  if (VALID_CATEGORIES.includes(c)) return c;
  return null;
}

function normalizeItem(raw) {
  const catEn = canonicalCategory(raw.category);
  if (!catEn) return null; // 过滤掉配饰等无关类别
  const seasonEn = ['Spring', 'Summer', 'Autumn', 'Winter', 'All'].includes(raw.season)
    ? raw.season
    : 'All';
  return {
    category: CATEGORY_CN[catEn] || catEn,
    color: toCn(COLOR_CN, raw.color, '未知'),
    season: SEASON_CN[seasonEn] || '四季',
    material: toCn(MATERIAL_CN, raw.material, '未知'),
    style: toCn(STYLE_CN, raw.style, '未知'),
    fit: toCn(FIT_CN, raw.fit, '未知'),
    pattern: toCn(PATTERN_CN, raw.pattern, '纯色'),
    brand: String(raw.brand || 'Unknown').trim() || 'Unknown',
    hasLogo: !!raw.hasLogo,
  };
}

// ---------- 去重：每个类别最多保留一件 ----------
// 无包围盒后无法用 IoU；改为按类别去重（保留最先出现的那件）。
function dedupItems(items) {
  const seen = new Set();
  const out = [];
  for (const item of items) {
    if (seen.has(item.category)) continue;
    seen.add(item.category);
    out.push(item);
  }
  return out;
}

async function callQwen(base64, mediaType) {
  const res = await fetch(`${CONFIG.QWEN_BASE_URL}/chat/completions`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${CONFIG.QWEN_API_KEY}`,
    },
    body: JSON.stringify({
      model: CONFIG.QWEN_MODEL,
      messages: [
        { role: 'system', content: SYSTEM_PROMPT },
        { role: 'user', content: buildUserMessage(base64, mediaType) },
      ],
      temperature: 0.1,
      max_tokens: 2000,
    }),
  });
  if (!res.ok) {
    const err = await res.text();
    throw new Error(`${CONFIG.QWEN_MODEL} - Qwen API ${res.status}: ${err}`);
  }
  const data = await res.json();
  const content = data?.choices?.[0]?.message?.content || '';
  return extractJSON(content);
}

// demo 模式：返回四大类各一件（仅用于跑通流程，无需 box）
function demoDetect() {
  const mk = (cat, extra) => ({
    category: cat,
    color: extra.color,
    season: extra.season,
    material: extra.material,
    style: extra.style,
    fit: extra.fit,
    pattern: extra.pattern,
    brand: 'Unknown',
    hasLogo: false,
  });
  return [
    mk('Top', {
      color: 'White', season: 'Summer', material: 'Cotton', style: 'Minimal', fit: 'Loose', pattern: 'Solid',
    }),
    mk('Bottom', {
      color: 'Blue', season: 'All', material: 'Denim', style: 'Casual', fit: 'Regular', pattern: 'Solid',
    }),
    mk('Shoes', {
      color: 'White', season: 'All', material: 'Leather', style: 'Sporty', fit: 'Regular', pattern: 'Solid',
    }),
    mk('Bag', {
      color: 'Brown', season: 'All', material: 'Leather', style: 'Formal', fit: 'Regular', pattern: 'Solid',
    }),
  ];
}

export async function detectClothing(imagePath) {
  const buf = fs.readFileSync(imagePath);
  const base64 = buf.toString('base64');
  const ext = imagePath.split('.').pop().toLowerCase();
  const mediaType =
    ext === 'png' ? 'image/png' : ext === 'webp' ? 'image/webp' : 'image/jpeg';

  let raw;
  if (CONFIG.DEMO_MODE) {
    raw = demoDetect();
  } else {
    raw = await callQwen(base64, mediaType);
  }
  // 1) 规范化 + 过滤非法类别；2) 按类别去重
  const normalized = raw.map((r) => normalizeItem(r)).filter(Boolean);
  return dedupItems(normalized);
}

// 导出内部辅助函数，便于单元测试
export { canonicalCategory, dedupItems, normalizeItem };

File: backend/qwen.py
"""
Qwen VL 服装识别模块。复刻原 lib/qwen.js 的 detectClothing + extractJson + normalize + dedup。
"""
import base64
import json
import logging
import os
import re
import time
from pathlib import Path

import requests

from config import API_KEY, DEMO_MODE, QWEN_MODEL, VISION_BASE_URL
from prompt import SYSTEM_PROMPT, build_user_message

logger = logging.getLogger("qwen")

# 受控英文词表（用于校正模型输出）
CATEGORY_VOCAB = {"Top", "Bottom", "Shoes", "Bag"}
CATEGORY_LABEL_ZH = {"Top": "上衣", "Bottom": "下装", "Shoes": "鞋", "Bag": "包"}

# 英文颜色 -> 中文颜色（覆盖常见模型输出）
COLOR_MAP_ZH = {
    "black": "黑", "white": "白", "gray": "灰", "grey": "灰", "beige": "米",
    "khaki": "卡其", "blue": "蓝", "navy": "藏蓝", "denim": "牛仔蓝", "red": "红",
    "green": "绿", "brown": "棕", "pink": "粉", "purple": "紫", "yellow": "黄",
    "orange": "橙", "cream": "米", "camel": "驼", "wine": "酒红", "maroon": "酒红",
    "olive": "橄榄绿", "mint": "薄荷绿", "skyblue": "天蓝", "lightblue": "浅蓝",
    "silver": "银", "gold": "金",
}

SEASON_VOCAB = {"春", "夏", "秋", "冬", "四季"}


def _media_type(path: str) -> str:
    ext = Path(path).suffix.lower().lstrip(".")
    mapping = {
        "jpg": "image/jpeg", "jpeg": "image/jpeg", "png": "image/png",
        "webp": "image/webp", "gif": "image/gif", "bmp": "image/bmp",
    }
    return mapping.get(ext, "image/jpeg")


def _extract_json(text: str):
    if text is None:
        return []
    s = text.strip()
    # 去掉 ```json ... ``` 代码块
    if s.startswith("```"):
        s = re.sub(r"^```[a-zA-Z]*\s*", "", s)
        s = re.sub(r"\s*```$", "", s)
        s = s.strip()
    # 截取首个 [ ... ] 片段
    start = s.find("[")
    end = s.rfind("]")
    if start != -1 and end != -1 and end > start:
        s = s[start : end + 1]
    try:
        data = json.loads(s)
    except Exception:
        return []
    if isinstance(data, dict):
        data = data.get("items", [])
    if not isinstance(data, list):
        return []
    return data


def _norm_category(cat: str) -> str:
    if not cat:
        return "Top"
    c = str(cat).strip().lower()
    # 直接命中
    for v in CATEGORY_VOCAB:
        if v.lower() == c:
            return v
    # 关键词兜底（覆盖模型可能输出的各种中文/英文表述）
    if any(k in c for k in ("dress", "连衣裙", "连身裙"," top", "shirt", "tee", "上衣", "外套", "卫衣", "t恤", "衫", "毛衣", "大衣", "西装", "夹克", "coat", "jacket", "sweater", "hoodie")):
        return "Top"
    if any(k in c for k in ("bottom", "pant", "trouser", "skirt", "下装", "裤子", "裤", "裙", "半裙" ,"裙装")):
        return "Bottom"
    if any(k in c for k in ("shoe", "sneaker", "boot", "鞋", "靴")):
        return "Shoes"
    if any(k in c for k in ("bag", "backpack", "包", "背包", "手提包", "挎包")):
        return "Bag"
    return "Top"


def _norm_color(color) -> str:
    if not color:
        return "未知"
    c = str(color).strip().lower()
    if c in COLOR_MAP_ZH:
        return COLOR_MAP_ZH[c]
    # 已经是中文则原样保留
    if re.search(r"[一-龥]", c):
        return str(color).strip()
    # 未知英文统一为“其他”
    return "其他"


def _norm_season(season) -> str:
    if not season:
        return "四季"
    s = str(season).strip()
    for v in SEASON_VOCAB:
        if v in s:
            return v
    return "四季"


def _norm_name(raw: dict, cat_zh: str) -> str:
    """生成单品名称：优先使用模型返回的 name，否则用属性兜底拼接。"""
    name = str(raw.get("name") or "").strip()
    if name:
        return name[:12]
    # 兜底：颜色 + 材质（非未知时） + 类别
    parts = []
    color = _norm_color(raw.get("color"))
    if color and color not in ("未知", "其他"):
        parts.append(color)
    material = str(raw.get("material") or "").strip()
    if material and material not in ("未知",):
        parts.append(material)
    parts.append(cat_zh)
    return "".join(parts)[:12]


def normalize(raw: dict) -> dict:
    cat_en = _norm_category(raw.get("category"))
    cat_zh = CATEGORY_LABEL_ZH.get(cat_en, cat_en)
    return {
        "category": cat_zh,
        "color": _norm_color(raw.get("color")),
        "season": _norm_season(raw.get("season")),
        "material": str(raw.get("material") or "未知").strip() or "未知",
        "style": str(raw.get("style") or "休闲").strip() or "休闲",
        "fit": str(raw.get("fit") or "常规").strip() or "常规",
        "pattern": str(raw.get("pattern") or "纯色").strip() or "纯色",
        "name": _norm_name(raw, cat_zh),
    }


def dedup(items: list) -> list:
    seen = set()
    out = []
    for it in items:
        key = (it["category"], it["color"])
        if key in seen:
            continue
        seen.add(key)
        out.append(it)
    return out


def detect_clothing(image_path: str) -> list:
    """识别单张图片中的所有服装单品，返回规范化后的 dict 列表。"""
    if DEMO_MODE:
        return []

    with open(image_path, "rb") as f:
        raw_data = f.read()
        b64 = base64.b64encode(raw_data).decode("ascii")

    img_size = len(raw_data)
    logger.info("开始 VL 分析图片: %s (%.1f KB)", image_path, img_size / 1024)

    payload = {
        "model": QWEN_MODEL,
        "messages": [
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": build_user_message(_media_type(image_path), b64)},
        ],
        "temperature": 0.2,
    }
    headers = {
        "Authorization": f"Bearer {API_KEY}",
        "Content-Type": "application/json",
    }
    url = VISION_BASE_URL.rstrip("/") + "/chat/completions"

    t0 = time.time()
    resp = requests.post(url, headers=headers, json=payload, timeout=120)
    elapsed = time.time() - t0
    logger.info("VL API 返回 status=%s, 耗时 %.1f 秒", resp.status_code, elapsed)

    if resp.status_code != 200:
        err_body = resp.text[:1000] if resp.text else "空响应"
        logger.error("VL API 请求失败 (status=%s), body=%s", resp.status_code, err_body)
        resp.raise_for_status()

    data = resp.json()
    logger.info("VL API 原始响应: %s", str(data)[:2000])
    content = data["choices"][0]["message"]["content"]
    logger.info("VL 返回内容: %s", content[:1000])
    raw_list = _extract_json(content)
    norm = [normalize(r) for r in raw_list]
    result = dedup(norm)
    logger.info("VL 识别结果: %d 个单品 (原始 %d 个)", len(result), len(raw_list))
    return result
